job-server-api/src/spark.jobserver/NamedRddSupport.scala

```diff
@@ -6,115 +6,191 @@ import java.util.concurrent.atomic.AtomicReference
+     * same time as getOrElseCreate() for the same name, and completes before the getOrElseCreate() call,
+     * then threads waiting for the result of getOrElseCreate() will unblock with the result of this
+     * update() call. When the getOrElseCreate() succeeds, it will replace the result of this update() call.
+     *
+     * @param name the unique name of the RDD. The uniqueness is scoped to the current SparkContext.
+     * @param rddGen a 0-ary function which will be called to generate the RDD in the caller's thread.
+     * @param forceComputation if true, forces the RDD to be computed by calling count().
+     * @param storageLevel the storage level to persist the RDD with. Default: StorageLevel.MEMORY_ONLY.
+     * @tparam T the generic type of the RDD.
+     * @return the RDD with the given name.
+     */
+    def update[T](name: String,
+                  rddGen: => RDD[T],
+                  forceComputation: Boolean = true,
+                  storageLevel: StorageLevel = defaultStorageLevel): RDD[T]
+
+    /**
+     * Destroys an RDD with the given name, if one existed. Has no effect if no RDD with this name exists.
+     *
+     * @param name the unique name of the RDD. The uniqueness is scoped to the current SparkContext.
+     */
+    def destroy(name: String): Unit
+
+    /**
+     * Returns the names of all named RDDs that are managed by the RddManager.
+     *
+     * Note: this returns a snapshot of RDD names at one point in time. The caller should always expect
+     * that the data returned from this method may be stale and incorrect.
+     *
+     * @param timeout if the RddManager doesn't respond within this timeout, an error will be thrown.
+     * @return a collection of RDD names representing RDDs managed by the RddManager.
+     */
+    def getNames()(implicit timeout: Timeout = defaultTimeout): Iterable[String]
+  }
+
+  private def rddPersister[T](forceComputation: Boolean) = new NamedObjectPersister[NamedRDD[T]] {
+    override def saveToContext(namedObj: NamedRDD[T], name: String) {
+      namedObj match {
+        case NamedRDD(rdd, storageLevel) =>
+          require(!forceComputation || storageLevel != StorageLevel.NONE,
+            "forceComputation implies storageLevel != NONE")
+          rdd.setName(name)
+          rdd.getStorageLevel match {
+            case StorageLevel.NONE => rdd.persist(storageLevel)
+            case currentLevel      => rdd.persist(currentLevel)
+          }
+          // TODO: figure out if there is a better way to force the RDD to be computed
+          if (forceComputation) rdd.count()
+      }
+    }
+
+//    override def getFromContext(c: ContextLike, name: String): NamedRDD[T] = {
+//      c.sparkContext.getPersistentRDDs.filter(_._2.name.equals(name)).map(_._2).headOption match {
+//        case Some(rdd: RDD[T]) => NamedRDD(rdd, rdd.getStorageLevel)
+//      }
+//    }
+
+    /**
+     * Calls rdd.persist(), which updates the RDD's cached timestamp, meaning it won't get
```

Comment:
I *think* this behaviour changed in spark, unless isLocallyCheckpointed is true persist() would not have any effect